### ISSUE TITLE
numericness failure in sorting

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9885]
Migrated from [rt.perl.org#69384](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=69384) (status was 'resolved')

Searchable as RT69384$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



$ perl \-lwe 'sub my\_cmp \{ $\_\[0\] \< $\_\[1\] ? "\-1" : $\_\[0\] == $\_\[1\] ? "0" : "1" \} print sort \{ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
Sort subroutine didn't return a numeric value at \-e line 1\.

Strings such as "\-1" are meant to be adequately numeric for Perl purposes\,
but sort is being picky\.  Can work around this by changing the comparator
block thus&#8203;:

$ perl \-lwe 'sub my\_cmp \{ $\_\[0\] \< $\_\[1\] ? "\-1" : $\_\[0\] == $\_\[1\] ? "0" : "1" \} print sort \{ 0 \+ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
12457

Can also work around it by side\-effect numerification&#8203;:

$ perl \-lwe 'sub my\_cmp \{ my $r = $\_\[0\] \< $\_\[1\] ? "\-1" : $\_\[0\] == $\_\[1\] ? "0" : "1"; \{ no warnings "void"; 0\+$r; \} return $r; \} print sort \{ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
12457

In the real application behind this bug report\, there is actually
good reason for the cmp function to generate its results as strings\.
I'll probably do side\-effect numerification as a permanent workaround for
the cmp function\.  Of course\, sort will still screw up when presented with
string\-based numbers generated by anything that's not built specifically
to work with sort\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.10.0:

Configured by Debian Project at Fri Aug 28 22:30:10 UTC 2009.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.26-2-amd64, archname=i486-linux-gnu-thread-multi
    uname='linux puccini 2.6.26-2-amd64 #1 smp fri aug 14 07:12:04 utc 2009 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.0 -Dsitearch=/usr/local/lib/perl/5.10.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=true, libperl=libperl.so.5.10.0
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.10.0:
    /etc/perl
    /usr/local/lib/perl/5.10.0
    /usr/local/share/perl/5.10.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.0:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From jpl@research.att.com

> $ perl \-lwe 'sub my\_cmp \{ $\_\[0\] \< $\_\[1\] ? "\-1" : $\_\[0\] == $\_\[1\] ? "0" : "1" \} print sort \{ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
> Sort subroutine didn't return a numeric value at \-e line 1\.

> Strings such as "\-1" are meant to be adequately numeric for
> Perl purposes\, but sort is being picky\.  Can work around
> this by changing the comparator block thus&#8203;:

> $ perl \-lwe 'sub my\_cmp \{ $\_\[0\] \< $\_\[1\] ? "\-1" : $\_\[0\] == $\_\[1\] ? "0" : "1" \} print sort \{ 0 \+ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
> 12457

> Can also work around it by side\-effect numerification&#8203;:

> $ perl \-lwe 'sub my\_cmp \{ my $r = $\_\[0\] \< $\_\[1\] ? "\-1" : $\_\[0\] == $\_\[1\] ? "0" : "1"; \{ no warnings "void"; 0\+$r; \} return $r; \} print sort \{ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
> 12457

> In the real application behind this bug report\, there is
> actually good reason for the cmp function to generate its
> results as strings\.  I'll probably do side\-effect
> numerification as a permanent workaround for the cmp
> function\.  Of course\, sort will still screw up when
> presented with string\-based numbers generated by anything
> that's not built specifically to work with sort\.

Sort is "screwing up" in the same way that overloading \+
with something that does subtraction will "screw up" a total\.

perldoc \-f sort

is pretty clear about wanting integer returns\.\.\.

&nbsp;    If SUBNAME is specified\, it gives the name of a subroutine
&nbsp;    that returns an integer less than\, equal to\, or greater
&nbsp;    than 0\, depending on how the elements of the list are to be
&nbsp;    ordered\.  \(The "\<=>" and "cmp" operators are extremely
&nbsp;    useful in such routines\.\)

\(note \*integer\*\, not "a string that looks like an integer"\)
and the error message\,

&nbsp;    Sort subroutine didn't return a numeric value at \-e line 1\.

leaves little to the imagination\.  It may well be that

&nbsp;    there is actually good reason for the \<application's>
&nbsp;    function to generate its results as strings

but calling it my\_cmp doesn't make it like the builtin cmp\.
Penalizing all the comparison routines that play by the rules
to accommodate some that don't doesn't sound like a good idea
to me\, particularly when the documentation\, error messages and
workarounds are this clear\.  \-\- jpl

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ikegami

On Fri\, Sep 25\, 2009 at 8&#8203;:21 AM\, Zefram \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> \# New Ticket Created by  Zefram
> \# Please include the string&#8203;:  \[perl \#69384\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=69384 >
>


> Can also work around it by side\-effect numerification&#8203;:
>
> $ perl \-lwe 'sub my\_cmp \{ my $r = $\_\[0\] \< $\_\[1\] ? "\-1" : $\_\[0\] == $\_\[1\] ?
> "0" : "1"; \{ no warnings "void"; 0\+$r; \} return $r; \} print sort \{
> my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
> 12457
>

Wouldn't
&nbsp;    \{ no warnings "void"; 0\+$r; \} return $r;
be simpler as
&nbsp;    return 0\+$r;

**Comment by p5pRT** [comment]
### From @davidnicol

On Fri\, Sep 25\, 2009 at 12&#8203;:38 PM\, John P\. Linderman
\<jpl@&#8203;research\.att\.com> wrote&#8203;:

> perldoc \-f sort
>
> is pretty clear about wanting integer returns\.\.\.
>
>    If SUBNAME is specified\, it gives the name of a subroutine
>    that returns an integer less than\, equal to\, or greater
>    than 0\, depending on how the elements of the list are to be
>    ordered\.  \(The "\<=>" and "cmp" operators are extremely
>    useful in such routines\.\)
>
> \(note \*integer\*\, not "a string that looks like an integer"\)
> and the error message\,


but this is Perl\! Programmers are supposed to be allowed to be
careless about switching between numbers and strings\.  How much of
"darkpan" is affected by this stricture? If Zefram were to deliver a
patch to\, for instance\, \`perform a "looks\-like\-a\-number" check and if
so\, try again\` in the spot where the exception in question is raised\,
would that patch have a chance of achieving application?

Ideally\, the "side effect numerification" would occur on the constant
string "1" in a COW manner\, even though the value has changed hands a
couple of times\, so that little smidge of CPU would only get used once
per constant instead of every time\, or the numerification could be
made persistent by adding the :lvalue attribute to the comparison
routine\.

**Comment by p5pRT** [comment]
### From @demerphq

2009/9/25 David Nicol \<davidnicol@&#8203;gmail\.com>&#8203;:
> On Fri\, Sep 25\, 2009 at 12&#8203;:38 PM\, John P\. Linderman
> \<jpl@&#8203;research\.att\.com> wrote&#8203;:
>
>> perldoc \-f sort
>>
>> is pretty clear about wanting integer returns\.\.\.
>>
>>    If SUBNAME is specified\, it gives the name of a subroutine
>>    that returns an integer less than\, equal to\, or greater
>>    than 0\, depending on how the elements of the list are to be
>>    ordered\.  \(The "\<=>" and "cmp" operators are extremely
>>    useful in such routines\.\)
>>
>> \(note \*integer\*\, not "a string that looks like an integer"\)
>> and the error message\,
>
>
> but this is Perl\! Programmers are supposed to be allowed to be
> careless about switching between numbers and strings\.  How much of
> "darkpan" is affected by this stricture? If Zefram were to deliver a
> patch to\, for instance\, \`perform a "looks\-like\-a\-number" check and if
> so\, try again\` in the spot where the exception in question is raised\,
> would that patch have a chance of achieving application?
>
> Ideally\, the "side effect numerification" would occur on the constant
> string "1" in a COW manner\, even though the value has changed hands a
> couple of times\, so that little smidge of CPU would only get used once
> per constant instead of every time\, or the numerification could be
> made persistent by adding the :lvalue attribute to the comparison
> routine\.
>

Its simple\, adding numification to the sort routine would slow the
routine down for /everybody/ for minimal benefit\.

So we break the general rule of thumb here\.

Seems pretty reasonable to me\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Eric Brine wrote&#8203;:
>Wouldn't
>    \{ no warnings "void"; 0\+$r; \} return $r;
>be simpler as
>    return 0\+$r;

No\, that would have different behaviour\.  I want to return a string\, for
string uses\, with its numeric value already populated to satisfy sort\.
Your version would return a pure numeric value\, losing the original
string\, so that if a string value is needed it will be generated anew
from the numeric value\.  In simple cases\, like the example I showed\,
the behaviour of these two will be identical\.  But number\-to\-string
conversion behaves badly around floating\-point zeroes\, and in general
there are many strings corresponding to a single numeric value\.

\-zefram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

>> On Fri\, Sep 25\, 2009 at 12&#8203;:38 PM\, John P\. Linderman
>> \<jpl@&#8203;research\.att\.com> wrote&#8203;:
>>> \(note \*integer\*\, not "a string that looks like an integer"\)

Apparently it doesn't need integer representation\, though&#8203;:

$ perl \-lwe 'sub my\_cmp \{ $\_\[0\] \< $\_\[1\] ? \-1\.0 : $\_\[0\] == $\_\[1\] ? 0\.0 : 1\.0 \} print sort \{ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
12457

Nor does its numeric value actually need to be an integer&#8203;:

$ perl \-lwe 'sub my\_cmp \{ $\_\[0\] \< $\_\[1\] ? \-1\.5 : $\_\[0\] == $\_\[1\] ? 0\.5 : 1\.5 \} print sort \{ my\_cmp\($a\, $b\) \} \(1\,5\,4\,7\,2\)'
12457

The unacceptability of strings violates the general statement in
perldata\(1\)&#8203;:

&nbsp;    A scalar may contain one single value in any of three different
&nbsp;    flavors&#8203;: a number\, a string\, or a reference\.  In general\,
&nbsp;    conversion from one form to another is transparent\.

This does indeed allow for conversion to be non\-transparent in some
situations\, but perlfunc\(1\) doesn't actually say that sort won't convert
from strings\.  It's rather surprising behaviour\, and if it's really
intended then the documentation needs to be much more explicit about
the limited conversion options\.

>2009/9/25 David Nicol \<davidnicol@&#8203;gmail\.com>&#8203;:
>> Ideally\, the "side effect numerification" would occur on the constant
>> string "1" in a COW manner\, even though the value has changed hands a
>> couple of times\, so that little smidge of CPU would only get used once
>> per constant instead of every time\,

I'll probably pre\-numerify the three string constants\, for the real
my\_cmp\(\)\.

demerphq wrote&#8203;:
>Its simple\, adding numification to the sort routine would slow the
>routine down for /everybody/ for minimal benefit\.

I wondered about that\.  It seemed like the only feasible reason for
intentionally having this behaviour\, but obviously there's already
enough logic to \*detect\* non\-numeric values\.  Putting conversion code
into that branch shouldn't affect the \(overwhelmingly common\) case
where the result already has an integer value stored\.  The only possible
performance\-relevant effect that I could see was from marginally increased
code size causing more cache misses\.  Good luck measuring that\.

But look at the code \(S\_sortcv in pp\_sort\.c\)\.  The numericness check
doesn't actually avoid having conversion logic present\, it takes the form
of a guard before the conversion\-capable SvIV\(\) invocation\.  It actually
\*increases\* both code size and the number of instructions performed in
the common case\.

It looks like the intent behind the check is not to reject "\-1"\, but
to reject "foo" and other such stuff\.  It seems to be trying to reject
the stuff that would generate an "Argument isn't numeric" warning when
used numerically\.  In some code I make an explicit check of whether
a value generates such a warning\, and die if it does\, but the process
is quite expensive\.  S\_sortcv seems to be trying to approximate that
behaviour by its cheap SvNIOKp\(\) check\.

I think it would be better to remove the SvNIOKp\(\) check entirely\,
and just go straight to SvIV\(\)\.  Let it generate the usual warning for
non\-numeric values\, judged by the ordinary Perlish criteria\.  We have
way too many subtly\-different sets of rules for different situations\.
Replacing this obscure criterion with a very commonly\-used one would
make the language more comprehensible and less surprising\.

On a quick test with blead\, just removing those two lines from S\_sortcv
gave precisely the behaviour that I'm proposing\.

\-zefram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Rafael Garcia\-Suarez wrote&#8203;:
>2009/9/26 Zefram \<zefram@&#8203;fysh\.org>&#8203;:
>> On a quick test with blead\, just removing those two lines from S\_sortcv
>> gave precisely the behaviour that I'm proposing\.
>
>Does it break any test ?

No\.  Apparently the current error case isn't being tested for\.  Attached
is a patch that makes the change in behaviour\, with corresponding test
and documentation\.

\-zefram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

<details><summary>Inline Patch</summary>

```diff
diff --git a/pod/perldiag.pod b/pod/perldiag.pod
index 1f7bc0b..026b68e 100644
--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -3920,12 +3920,6 @@ the smart match.
 (F) An ancient error message that almost nobody ever runs into anymore.
 But before sort was a keyword, people sometimes used it as a filehandle.
 
-=item Sort subroutine didn't return a numeric value
-
-(F) A sort comparison routine must return a number.  You probably blew
-it by not using C<< <=> >> or C<cmp>, or by not using them correctly.
-See L<perlfunc/sort>.
-
 =item Sort subroutine didn't return single value
 
 (F) A sort comparison subroutine may not return a list value with more
diff --git a/pp_sort.c b/pp_sort.c
index a27232f..4abec80 100644
--- a/pp_sort.c
+++ b/pp_sort.c
@@ -1757,8 +1757,6 @@ S_sortcv(pTHX_ SV *const a, SV *const b)
     CALLRUNOPS(aTHX);
     if (PL_stack_sp != PL_stack_base + 1)
 	Perl_croak(aTHX_ "Sort subroutine didn't return single value");
-    if (!SvNIOKp(*PL_stack_sp))
-	Perl_croak(aTHX_ "Sort subroutine didn't return a numeric value");
     result = SvIV(*PL_stack_sp);
     while (PL_scopestack_ix > oldscopeix) {
 	LEAVE;
@@ -1799,8 +1797,6 @@ S_sortcv_stacked(pTHX_ SV *const a, SV *const b)
     CALLRUNOPS(aTHX);
     if (PL_stack_sp != PL_stack_base + 1)
 	Perl_croak(aTHX_ "Sort subroutine didn't return single value");
-    if (!SvNIOKp(*PL_stack_sp))
-	Perl_croak(aTHX_ "Sort subroutine didn't return a numeric value");
     result = SvIV(*PL_stack_sp);
     while (PL_scopestack_ix > oldscopeix) {
 	LEAVE;
@@ -1829,8 +1825,6 @@ S_sortcv_xsub(pTHX_ SV *const a, SV *const b)
     (void)(*CvXSUB(cv))(aTHX_ cv);
     if (PL_stack_sp != PL_stack_base + 1)
 	Perl_croak(aTHX_ "Sort subroutine didn't return single value");
-    if (!SvNIOKp(*PL_stack_sp))
-	Perl_croak(aTHX_ "Sort subroutine didn't return a numeric value");
     result = SvIV(*PL_stack_sp);
     while (PL_scopestack_ix > oldscopeix) {
 	LEAVE;
diff --git a/t/base/cond.t b/t/base/cond.t
diff --git a/t/op/sort.t b/t/op/sort.t
index 616761a..8484827 100644
--- a/t/op/sort.t
+++ b/t/op/sort.t
@@ -6,7 +6,7 @@ BEGIN {
     require 'test.pl';
 }
 use warnings;
-plan( tests => 144 );
+plan( tests => 146 );
 
 # these shouldn't hang
 {
@@ -801,3 +801,10 @@ is("@b", "10 9 8 7 6 5 4 3 2 1", "return with SVs on stack");
 sub ret_with_stacked { $_ = ($a<=>$b) + do {return $b <=> $a} }
 @b = sort ret_with_stacked 1..10;
 is("@b", "10 9 8 7 6 5 4 3 2 1", "return with SVs on stack");
+
+# Comparison code should be able to give result in non-integer representation.
+sub cmp_as_string($$) { $_[0] < $_[1] ? "-1" : $_[0] == $_[1] ? "0" : "+1" }
+@b = sort { cmp_as_string($a, $b) } (1,5,4,7,3,2,3);
+is("@b", "1 2 3 3 4 5 7", "comparison result as string");
+@b = sort cmp_as_string (1,5,4,7,3,2,3);
+is("@b", "1 2 3 3 4 5 7", "comparison result as string");
diff --git a/utils/Makefile b/utils/Makefile
```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

2009/9/26 Zefram \<zefram@&#8203;fysh\.org>&#8203;:
> demerphq wrote&#8203;:
>>Its simple\, adding numification to the sort routine would slow the
>>routine down for /everybody/ for minimal benefit\.
>
> I wondered about that\.  It seemed like the only feasible reason for
> intentionally having this behaviour\, but obviously there's already
> enough logic to \*detect\* non\-numeric values\.  Putting conversion code
> into that branch shouldn't affect the \(overwhelmingly common\) case
> where the result already has an integer value stored\.  The only possible
> performance\-relevant effect that I could see was from marginally increased
> code size causing more cache misses\.  Good luck measuring that\.

I tend to agree : an element will get compared many times\, but converted
only once\.

> On a quick test with blead\, just removing those two lines from S\_sortcv
> gave precisely the behaviour that I'm proposing\.

Does it break any test ?

**Comment by p5pRT** [comment]
### From @ikegami

On Sat\, Sep 26\, 2009 at 8&#8203;:47 AM\, Zefram \<zefram@&#8203;fysh\.org> wrote&#8203;:

> Eric Brine wrote&#8203;:
> >Wouldn't
> >    \{ no warnings "void"; 0\+$r; \} return $r;
> >be simpler as
> >    return 0\+$r;
>
> No\, that would have different behaviour\.  I want to return a string\, for
> string uses\, with its numeric value already populated to satisfy sort\.
> Your version would return a pure numeric value\, losing the original
> string\, so that if a string value is needed it will be generated anew
> from the numeric value\.  In simple cases\, like the example I showed\,
> the behaviour of these two will be identical\.  But number\-to\-string
> conversion behaves badly around floating\-point zeroes\, and in general
> there are many strings corresponding to a single numeric value\.
>

I know\, but we're talking about the about the return of a comparison
function\. Why would you need that as a string\, even if you used it for
something other than sort\.

And if there is such a need\, then we should have cmp and \<=> return
dualvars\.

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Eric Brine wrote&#8203;:
>I know\, but we're talking about the about the return of a comparison
>function\. Why would you need that as a string\, even if you used it for
>something other than sort\.

The intention is that the same comparison function can be used in more
than one way\.  It's not \*only\* for use with sort\.  It's intended to
be used with sort\, but also for the return values to have a strictly
guaranteed string form for use in string context\.

The decision to return the values in string form was marginal\.
The number\-to\-string conversions \*should\* always actually produce the
strings that I want\, but there's nastiness around numeric zeroes\, and
I'm not comfortable about relying on the long\-term stability of all the
relevant semantics\.  The string form is of primary importance to the
module\, and the string\-to\-number conversion is much more reliable in
the IV range \(or so I thought\)\.  The module is Math&#8203;::Decimal\, and the
function is dec\_cmp\(\)\, if you want to examine the context\.

So\, it's definitely an oddball situation\.  I made the comparison function
dual\-use to fit the particular style and conventions of that module\.
I wouldn't try the equivalent with most arithmetic modules\.  This is
presumably why no one has run into this problem before\.

\-zefram

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgarcia

2009/9/26 Zefram \<zefram@&#8203;fysh\.org>&#8203;:
> Rafael Garcia\-Suarez wrote&#8203;:
>>2009/9/26 Zefram \<zefram@&#8203;fysh\.org>&#8203;:
>>> On a quick test with blead\, just removing those two lines from S\_sortcv
>>> gave precisely the behaviour that I'm proposing\.
>>
>>Does it break any test ?
>
> No\.  Apparently the current error case isn't being tested for\.  Attached
> is a patch that makes the change in behaviour\, with corresponding test
> and documentation\.

Thanks\, applied as 93e19c0f634ff075d96380394c4e3449389368d8

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

Citeren Zefram \<zefram@&#8203;fysh\.org>&#8203;:

> Rafael Garcia\-Suarez wrote&#8203;:
>> 2009/9/26 Zefram \<zefram@&#8203;fysh\.org>&#8203;:
>>> On a quick test with blead\, just removing those two lines from S\_sortcv
>>> gave precisely the behaviour that I'm proposing\.
>>
>> Does it break any test ?
>
> No\.  Apparently the current error case isn't being tested for\.  Attached
> is a patch that makes the change in behaviour\, with corresponding test
> and documentation\.
>

Do note that this patched altered the behaviour of sort when the sort  
function returned a string that does not look like a number\.

It should be reconsidered if these changes were intended&#8203;:

First change&#8203;:

'Sort subroutine didn't return a numeric value' was a fatal error;
'Argument %s isn't numeric' is a warning;

Second change&#8203;:

The description of the error/warning; the warning now says that the  
string isn't a number but it doesn't say why it should be a number\.


\#\!/usr/bin/perl \-l

use strict;
use warnings;

my @&#8203;list = \(1 \.\. 20\);
my @&#8203;list2 = sort \{ print "in sort \($a\, $b\)"; "foo"; \} @&#8203;list;
print join\(" & "\, @&#8203;list2\);
print "done";
\_\_END\_\_

$ perl\-5\.6\.0 /tmp/rt\-69384\.pl
in sort \(1\, 2\)
Sort subroutine didn't return a numeric value at /tmp/rt\-69384\.pl line 7\.

$ perl\-5\.10\.0 /tmp/rt\-69384\.pl
in sort \(1\, 2\)
Sort subroutine didn't return a numeric value at /tmp/rt\-69384\.pl line 7\.

$ perl\-5\.11\.0 /tmp/rt\-69384\.pl
in sort \(1\, 2\)
Argument "foo" isn't numeric at /tmp/rt\-69384\.pl line 7\.
in sort \(3\, 4\)
in sort \(5\, 6\)
in sort \(7\, 8\)
in sort \(9\, 10\)
in sort \(11\, 12\)
in sort \(13\, 14\)
in sort \(15\, 16\)
in sort \(17\, 18\)
in sort \(19\, 20\)
in sort \(16\, 17\)
in sort \(14\, 15\)
in sort \(12\, 13\)
in sort \(10\, 11\)
in sort \(8\, 9\)
in sort \(6\, 7\)
in sort \(4\, 5\)
in sort \(2\, 3\)
in sort \(18\, 19\)
1 & 2 & 3 & 4 & 5 & 6 & 7 & 8 & 9 & 10 & 11 & 12 & 13 & 14 & 15 & 16 &  
17 & 18 & 19 & 20
done


IMHO&#8203;: the fatal error should remain when the string does not look like  
a number\.

That is&#8203;:

sort \{ "foo" \} \# => fatal
sort \{ "\-1" \}  \# => ok


Best regards\,

Bram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Bram wrote&#8203;:
>'Sort subroutine didn't return a numeric value' was a fatal error;
>'Argument %s isn't numeric' is a warning;

I was mindful of this change\.  As I said\, I think we're better off with
the semantics being consistent with the rest of perl\, and that entails
allowing "foo" to be used as a number with only an optional warning\.
I'd be all in favour of a pragma to make that warning \(locally\) fatal\,
which should apply not only to sort but also to anywhere else that
currently generates the warning\.  \(In fact\, can't "warnings" already do
this somehow?  It's not well documented\.\)

>The description of the error/warning; the warning now says that the  
>string isn't a number but it doesn't say why it should be a number\.

Ah\, that's unfortunate\.  \(0\+"foo"\) says "\.\.\. isn't numeric in addition
\(\+\)"\.  It would be nice to get the equivalent for sort\.

\-zefram
